Core/clim-basic/utils.lisp

```diff
@@ -808,161 +808,76 @@ a flag CLOSED is T then beginning and end of the list are consecutive too."
-                               (current-break (if-let ((pos (skip-whitespace string opportunity)))
-                                                pos
-                                                (return-from breaks-rec nil))))
-                           (multiple-value-bind (breaks remaining-cost)
-                               (breaks-rec 0 current-break end (1+ i))
-                             (when (or (null best-cost)
-                                       (< (+ current-cost remaining-cost) best-cost))
-                               (setf best-breaks (list* current-break breaks)
-                                     best-cost (+ current-cost remaining-cost))))))
-                        ((null best-cost)
-                         (return-from breaks-rec
-                           (if (> offset 0)
-                               (multiple-value-bind (breaks remaining-cost)
-                                   (breaks-rec 0 start end i)
-                                 (return-from breaks-rec
-                                   (values (list* start breaks)
-                                           (+ remaining-cost current-margin))))
-                               (let* ((char-breaks (%line-breaks-1 string width
-                                                                   offset margin
-                                                                   start opportunity))
-                                      (new-start (alexandria:last-elt char-breaks)))
-                                 (multiple-value-bind (breaks remaining-cost)
-                                     (breaks-rec 0 new-start end i)
-                                   (values (append char-breaks breaks) remaining-cost))))))
-                        (t #1=(return-from breaks-rec
-                                (values best-breaks best-cost))))
-                finally #1#)))
-    (breaks-rec initial-offset start end 0)))
-
-;;; Implementing line breaking as defined in Unicode[1] is left as an
-;;; excercise for the reader. When implemented it should be wired to
-;;; the strategy designated by T. Current approach break on space, but
-;;; in "real" languages lines may /need to/ or /must to/ be split under
-;;; various conditions - hyphen, hard space, infix numeric separators
-;;; etc. Optimally this should be implemented in cl-unicode and used
-;;; from there.  [1] https://unicode.org/reports/tr14/ -- jd 2019-01-08
-(defun line-breaks (string width &key (break-strategy t) initial-offset margin (start 0) end)
-  "Function takes a string and returns a list of indexes where it should be split.
+;;; Returns an index of the character _before_ which we should break
+;;; the line (may be used as END argument in SUBSEQ). When FROM-END is
+;;; T, then we return the rightmost opportunity which does not violate
+;;; hard line breaks. Second value indicates the kind of a line break.
```

Comment:
```
foo bar\nqux bam
   ^   ^^    ^
```

In this string we have three line break opportunities, between foo and bar, between bar and qux and between qux and bam. Rightmost break opportunity is between qux and bam, but newline is a hard line break, so we pick that. (otherwise we would not break on a mandatory break character). Previous version of the code assumed, that there are no newlines and such in the text, this version is more robust (and in the future i'd like to use it earlier than in seos-write-string, namely in stream-write-string during line breaks).